vowpalwabbit/ftrl.cc

```diff
@@ -434,17 +436,25 @@ base_learner* ftrl_setup(VW::setup_base_i& stack_builder)
 
   learner<ftrl, example>* l;
   if (all.audit || all.hash_inv)
-    l = &init_learner(b, learn_ptr, predict<true>, UINT64_ONE << all.weights.stride_shift(),
-        stack_builder.get_setupfn_name(ftrl_setup) + "-" + algorithm_name + "-audit");
-  else
-    l = &init_learner(b, learn_ptr, predict<false>, UINT64_ONE << all.weights.stride_shift(),
-        stack_builder.get_setupfn_name(ftrl_setup) + "-" + algorithm_name, learn_returns_prediction);
-  l->set_sensitivity(sensitivity);
-  if (all.audit || all.hash_inv)
-    l->set_multipredict(multipredict<true>);
+    l = VW::LEARNER::make_base_learner(std::move(b), learn_ptr, predict<true>,
+        stack_builder.get_setupfn_name(ftrl_setup) + "-" + algorithm_name + "-audit", prediction_type_t::scalar,
+        label_type_t::simple)
+            .set_params_per_weight(UINT64_ONE << all.weights.stride_shift())
+            .set_sensitivity(sensitivity)
+            .set_multipredict(multipredict<true>)
+            .set_save_load(save_load)
+            .set_end_pass(end_pass)
+            .build();
   else
-    l->set_multipredict(multipredict<false>);
-  l->set_save_load(save_load);
-  l->set_end_pass(end_pass);
+    l = VW::LEARNER::make_base_learner(std::move(b), learn_ptr, predict<false>,
+        stack_builder.get_setupfn_name(ftrl_setup) + "-" + algorithm_name, prediction_type_t::scalar,
+        label_type_t::simple)
+            .set_learn_returns_prediction(learn_returns_prediction)
+            .set_params_per_weight(UINT64_ONE << all.weights.stride_shift())
+            .set_sensitivity(sensitivity)
+            .set_multipredict(multipredict<false>)
+            .set_save_load(save_load)
+            .set_end_pass(end_pass)
+            .build();
   return make_base(*l);
```

Comment:
There's a lot of duplication here. Can you do the shared things outside of the `if`?

Reply:
Hm yeah I see, thats pretty cumbersome isnt it

Reply:
Suggestion:
```c++
... predict_ptr = (all.audit || all.hash_inv) ? predict<true> : predict<false>;
... multipredict_ptr = (all.audit || all.hash_inv) ? multipredict<true> : multipredict<false>;

l = VW::LEARNER::make_base_learner(std::move(b), learn_ptr, predict_ptr,
        stack_builder.get_setupfn_name(ftrl_setup) + "-" + algorithm_name + "-audit", prediction_type_t::scalar,
        label_type_t::simple)
            .set_learn_returns_prediction(learn_returns_prediction)
            .set_params_per_weight(UINT64_ONE << all.weights.stride_shift())
            .set_sensitivity(sensitivity)
            .set_multipredict(multipredict_ptr)
            .set_save_load(save_load)
            .set_end_pass(end_pass)
            .build();
```

Reply:
Also added conditional for name